File: backend/app/api/routes/procurement_history_items.py
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.api.validation import get_or_404
from app.models.procurement_history_item import ProcurementHistoryItem
from app.schemas.procurement_history_item import ProcurementHistoryItemRead

router = APIRouter()


@router.get("", response_model=list[ProcurementHistoryItemRead])
def list_procurement_history_items(
    skip: int = 0,
    limit: int = 100,
    company_id: UUID | None = None,
    category: str | None = None,
    item_name: str | None = None,
    country: str | None = None,
    supplier_name: str | None = None,
    purchased_from: date | None = None,
    purchased_to: date | None = None,
    db: Session = Depends(get_db),
) -> list[ProcurementHistoryItem]:
    query = select(ProcurementHistoryItem)
    if company_id:
        query = query.where(ProcurementHistoryItem.company_id == company_id)
    if category:
        query = query.where(ProcurementHistoryItem.category == category)
    if item_name:
        query = query.where(ProcurementHistoryItem.item_name.ilike(f"%{item_name}%"))
    if country:
        query = query.where(ProcurementHistoryItem.supplier_country == country)
    if supplier_name:
        query = query.where(ProcurementHistoryItem.supplier_name.ilike(f"%{supplier_name}%"))
    if purchased_from:
        query = query.where(ProcurementHistoryItem.purchased_at >= purchased_from)
    if purchased_to:
        query = query.where(ProcurementHistoryItem.purchased_at <= purchased_to)
    return list(db.scalars(query.offset(skip).limit(limit)).all())


@router.get("/{procurement_history_item_id}", response_model=ProcurementHistoryItemRead)
def get_procurement_history_item(
    procurement_history_item_id: UUID,
    db: Session = Depends(get_db),
) -> ProcurementHistoryItem:
    return get_or_404(db, ProcurementHistoryItem, procurement_history_item_id, "Procurement history item")

File: frontend/lib/api/procurement-history-items.ts
import { apiGet } from "@/lib/api-client";

export type ProcurementHistoryItemSummary = {
  id: string;
  company_id: string;
  supplier_name?: string | null;
  supplier_country?: string | null;
  item_name: string;
  category?: string | null;
  sku?: string | null;
  quantity?: string | null;
  unit?: string | null;
  unit_price?: string | null;
  currency?: string | null;
  lead_time_weeks?: string | null;
  quality_rating?: string | null;
  price_assessment?: string | null;
  improvement_potential?: string | null;
  purchased_at?: string | null;
  source_document?: string | null;
  notes?: string | null;
  metadata_json?: Record<string, unknown>;
  created_at?: string;
  updated_at?: string;
};

export type ProcurementHistoryItemRead = ProcurementHistoryItemSummary;

export type ProcurementHistoryItemListFilters = {
  company_id?: string;
  category?: string;
  item_name?: string;
  country?: string;
  supplier_name?: string;
  purchased_from?: string;
  purchased_to?: string;
  skip?: number;
  limit?: number;
};

export function listProcurementHistoryItems(filters?: ProcurementHistoryItemListFilters) {
  return apiGet<ProcurementHistoryItemSummary[]>("/api/procurement-history-items", { query: filters, cache: "no-store" });
}

export function getProcurementHistoryItem(id: string) {
  return apiGet<ProcurementHistoryItemRead>(`/api/procurement-history-items/${id}`, { cache: "no-store" });
}
